### DISCUSSION TITLE
Unable to load FIPS provider programmatically

### DISCUSSION THREAD

**Discussion by AustinMayerhofer** [openssl/openssl#23665] [Q&A] [answered]
Hi all,

I have my main.c file taken from https://www.openssl.org/docs/manmaster/man7/fips_module.html:
```
#include <stdio.h>
#include <stdlib.h>
#include <openssl/provider.h>

int main() {
    OSSL_PROVIDER *fips;
    OSSL_PROVIDER *base;

    fips = OSSL_PROVIDER_load(NULL, "fips");
    if (fips == NULL) {
        printf("Failed to load FIPS provider\n");
        exit(EXIT_FAILURE);
    }
    base = OSSL_PROVIDER_load(NULL, "base");
    if (base == NULL) {
        OSSL_PROVIDER_unload(fips);
        printf("Failed to load base provider\n");
        exit(EXIT_FAILURE);
    }

    /* Rest of application */

    OSSL_PROVIDER_unload(base);
    OSSL_PROVIDER_unload(fips);
    exit(EXIT_SUCCESS);
}

```

and my Makefile:

```
all: main.c
	gcc -g -Wall -Werror libcrypto.dylib main.c \
	-I/<path...to...>/openssl/include \
	-o main
	OPENSSL_MODULES=/<path...to...>/openssl/providers ./main
```

Yet `main` returns "Failed to load FIPS provider", with or without setting OPENSSL_MODULES.

Where does the application look for the FIPS provider when calling OSSL_PROVIDER_load()? I followed the example at https://www.youtube.com/watch?v=geAtEXbHaFg of running `./util/wrap.pl -fips .openssl/bin/openssl list -provider-path /Users/austinmayerhofer/Desktop/Programming/openssl/providers -providers` which did show the FIPS provider, but I'm not sure how this openssl executable relates to the libcrypto library I'm linking to my app.

Any help is much appreciated, thanks!

**Comment by mattcaswell** [comment, marked as answer]
Did you run through the whole video? That command just tells you that you built of a version of OpenSSL that has the fips provider? You need to actually install it and configure it. The rest of that video runs you through the process.

You might also read this file:

https://github.com/openssl/openssl/blob/master/README-FIPS.md
